### ISSUE TITLE
(Question) - how to import workerpool in es6 module

### ISSUE THREAD

**Issue by GOVYANSONG** [josdejong/workerpool#43]
Our environment is typescript, es6 module and webpack. When I tried to compile my project, i received following webpack errors:

WARNING in ./node_modules/workerpool/lib/environment.js
    2:21-28 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
     @ ./node_modules/workerpool/lib/environment.js
     @ ./node_modules/workerpool/index.js
     @ ./ClientApp/util/Fetcher.ts
     @ ./ClientApp/util/index.tsx
     @ ./ClientApp/boot-client.tsx
    
    WARNING in ./node_modules/workerpool/lib/WorkerHandler.js
    8:21-28 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
     @ ./node_modules/workerpool/lib/WorkerHandler.js
     @ ./node_modules/workerpool/lib/Pool.js
     @ ./node_modules/workerpool/index.js
     @ ./ClientApp/util/Fetcher.ts
     @ ./ClientApp/util/index.tsx
     @ ./ClientApp/boot-client.tsx

I guess my question is how to import workerpool in es6 module. or is it possible.

**Comment by josdejong** [comment]
The [`environment.js`](https://github.com/josdejong/workerpool/blob/master/lib/environment.js) file contains some commonjs tricks to prevent node.js specific code being bundled in the browser bundle. I think there is a refactoring required to make this work with ES6, we probably have to create two separate entry points (one for browser, one for node.js) or something like that. Anyone interested into looking into this?

I don't know if there are ways to exclude the library from TypeScript checking somehow. Webpack itself is capable of compiling the code so there may be a way to get this working as is.

**Comment by GOVYANSONG** [comment]
Jos, thanks for the timely response. Very appreciated. After experimented with few twits in webpack in vain, we quickly explored other projects. Here is the one we are investigating: threads.js. It supports es6 and backward compatible besides other desirable features.

**Comment by knnhcn** [comment]
I am getting a similar warning in Angular 6:

```
WARNING in ./node_modules/workerpool/lib/environment.js
2:21-28 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

WARNING in ./node_modules/workerpool/lib/WorkerHandler.js
8:21-28 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
Compiled with warnings.
```

**Comment by Holgrabus** [comment]
I'm trying to use this module with Meteor and i got an error with imports. Is there a way to do it ?

**Error :**
```
(function (exports, require, module, __filename, __dirname) { import { Meteor } from 'meteor/meteor';
                                                              ^^^^^^

 SyntaxError: Unexpected token import
```

**Simple example :**

server\main.js
```
workerpool = require('workerpool');
workers = workerpool.pool(appConfig.directories.app + "imports/api/workers.js", {"minWorkers":1,"maxWorkers":3});
```

server\main.js
```
import {Clients} from "./clients";

function test(clientId) {
    return Clients.findOne({id: clientId}).name;
}
```

**Comment by josdejong** [comment]
I will try to find some time to address this issue coming weeks. I have the feeling that it should be relatively easy to fix and it is absolutely necessary to fix this in order for this library to stay relevant. 

If there is someone else who can work on it please let me know.

**Comment by josdejong** [comment]
I've found a nice solution to make `workerpool` work in ES6+Webpack projects. It requires removing some special tricks that are there to make the code work with AMD. I'm not sure if AMD is still used anywhere, but it may be acceptable to drop support for AMD in order to get this work in the most common environment (ES5+Webpack). Still trying to get it work in both...

**Comment by josdejong** [comment]
@robhawkes  do you think AMD support is still relevant today (see #5)?

**Comment by robhawkes** [comment]
No complaints from me to remove AMD support :)

**Comment by josdejong** [comment]
:+1:

**Comment by Holgrabus** [comment]
I agree, but my opinion is not really important i think ^^

**Comment by josdejong** [comment]
I've just released `workerpool@3.0.0` which works with ES6+Webpack, and drops support for AMD.

**Comment by pyxarez** [comment]
![image](https://user-images.githubusercontent.com/23152279/56135433-66bfed80-5f99-11e9-99c0-61663738bd1d.png)
@josdejong , Hello! I've upgraded to 3.0.1 version recently and see the following errors while importing like that 
![image](https://user-images.githubusercontent.com/23152279/56135973-7a1f8880-5f9a-11e9-9074-856d4301d6bd.png), can you please tell me what I'm doing wrong?
webpack: 4
typescript: 3

**Comment by pyxarez** [comment]
![image](https://user-images.githubusercontent.com/23152279/56135433-66bfed80-5f99-11e9-99c0-61663738bd1d.png)
@josdejong , Hello! I've upgraded to 3.0.1 version recently and see the following errors while importing like that 
![image](https://user-images.githubusercontent.com/23152279/56135973-7a1f8880-5f9a-11e9-9074-856d4301d6bd.png), can you please tell me what I'm doing wrong?
webpack: 4
typescript: 3

**Comment by josdejong** [comment]
@pyxarez here two examples of using workerpool in an ES6 project:

React: https://codesandbox.io/s/xz9jwnyvrw
React+ts: https://codesandbox.io/s/50lvkj9wwk

There should be no need to refer to import from `workerpool/dist/workerpool`, `workerpool` should just work.

Does that help?

**Comment by pyxarez** [comment]
@josdejong I made this changes 
![image](https://user-images.githubusercontent.com/23152279/58665716-28f50980-833a-11e9-95e9-6180aadc3bff.png)
Now it works!

**Comment by josdejong** [comment]
:+1: good to hear Alexander!

**Comment by LesterLyu** [comment]
React webpack is giving warnings, but it works!: 
```
Compiled with warnings.

./node_modules/workerpool/lib/WorkerHandler.js
Critical dependency: the request of a dependency is an expression
```
It seems webpack does not like the `tryRequire` function in `WorkerHandler.js`.

**Comment by josdejong** [comment]
That's an interesting case @LesterLyu . I think your proposal of replacing the generic `tryRequire` specifi for workerThreads makes sense. I've addressed this via https://github.com/josdejong/workerpool/commit/b38c515da2b052efaa7e4f018a67892bfe8bc366 (and https://github.com/josdejong/workerpool/commit/9c555644e1fed333e6e894922d4535e624342f62)

**Comment by RishabhSharmaGit** [comment]
> @josdejong I made this changes to webpack configurations
> ![image](https://user-images.githubusercontent.com/23152279/58665716-28f50980-833a-11e9-95e9-6180aadc3bff.png)
> Now it works!

Thanks a lot @pyxarez , your solution helped me in running workerpool with typescript. Though I didn't understand why it worked :P

**Comment by josdejong** [comment]
Will reopen this. Ideally, it shouldn't be needed to do these kind of configuration yourself. We'll have to investigate and see if we can fix this in `workerpool` itself.

**Comment by cawa-93** [comment]
Maybe it makes sense to compile several separate collections?
In the context of browser execution, the file size is important
It would be great to have a separate compilation created exclusively for the browser (any code that is required for the server environment is deleted).

**Comment by josdejong** [comment]
Yes I think so.

**Comment by loganpowell** [comment]
is there an example webpack setup?

**Comment by josdejong** [comment]
I've just published `workerpool@4.0.0` which should solve the issues with optional external dependencies. Please give it a try.

**Comment by josdejong** [comment]
FYI: browser, node.js, and Webpack now use the same bundle in v4.0.0. I use a trick inspired by [require-fool-webpack](https://github.com/sindresorhus/require-fool-webpack) to make Webpack ignore the node.js specific dependencies that caused issues.
